cpp/src/io/utilities/datasource.cpp

```diff
@@ -199,6 +200,67 @@ class direct_read_source : public file_source {
   }
 };
 
+/**
+ * @brief Implementation class for reading from a device buffer source
+ */
+class device_buffer_source final : public datasource {
+ public:
+  explicit device_buffer_source(cudf::device_span<std::byte const> d_buffer) : _d_buffer{d_buffer}
+  {
+  }
+
+  size_t host_read(size_t offset, size_t size, uint8_t* dst) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpy(dst, _d_buffer.data() + offset, count, cudaMemcpyDeviceToHost));
+    return count;
+  }
+
+  std::unique_ptr<buffer> host_read(size_t offset, size_t size) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    std::vector<std::byte> h_data(count);
+    CUDF_CUDA_TRY(
+      cudaMemcpy(h_data.data(), _d_buffer.data() + offset, count, cudaMemcpyDeviceToHost));
+    return std::make_unique<owning_buffer<std::vector<std::byte>>>(std::move(h_data));
+  }
+
+  [[nodiscard]] bool supports_device_read() const override { return true; }
+
+  std::future<size_t> device_read_async(size_t offset,
+                                        size_t size,
+                                        uint8_t* dst,
+                                        rmm::cuda_stream_view stream) override
+  {
+    auto const count = std::min(size, this->size() - offset);
+    CUDF_CUDA_TRY(cudaMemcpyAsync(
+      dst, _d_buffer.data() + offset, count, cudaMemcpyDeviceToDevice, stream.value()));
+    return std::async(std::launch::async, [count] { return count; });
```

Comment:
If I understand `std::async` correctly, we might be better off with `std::launch::deferred` here, to avoid spawning a new thread. Since the work in the `std::async` is trivial, we should be fine just deferring it to `future.get()` call time.
```suggestion
    return std::async(std::launch::deferred, [count] { return count; });
```